crates/bevy_ecs/src/reflect.rs

```diff
@@ -42,17 +42,21 @@ pub struct ReflectComponent(ReflectComponentFns);
 #[derive(Clone)]
 pub struct ReflectComponentFns {
     /// Function pointer implementing [`ReflectComponent::insert()`].
-    pub insert: fn(&mut World, Entity, &dyn Reflect),
+    pub insert: fn(&mut EntityMut, &dyn Reflect),
     /// Function pointer implementing [`ReflectComponent::apply()`].
-    pub apply: fn(&mut World, Entity, &dyn Reflect),
+    pub apply: fn(&mut EntityMut, &dyn Reflect),
     /// Function pointer implementing [`ReflectComponent::apply_or_insert()`].
-    pub apply_or_insert: fn(&mut World, Entity, &dyn Reflect),
+    pub apply_or_insert: fn(&mut EntityMut, &dyn Reflect),
     /// Function pointer implementing [`ReflectComponent::remove()`].
-    pub remove: fn(&mut World, Entity),
+    pub remove: fn(&mut EntityMut),
+    /// Function pointer implementing [`ReflectComponent::contains()`].
+    pub contains: fn(EntityRef) -> bool,
     /// Function pointer implementing [`ReflectComponent::reflect()`].
-    pub reflect: fn(&World, Entity) -> Option<&dyn Reflect>,
+    pub reflect: fn(EntityRef) -> Option<&dyn Reflect>,
     /// Function pointer implementing [`ReflectComponent::reflect_mut()`].
-    pub reflect_mut: unsafe fn(&World, Entity) -> Option<Mut<dyn Reflect>>,
+    pub reflect_mut: for<'a> fn(&'a mut EntityMut<'a>) -> Option<Mut<'a, dyn Reflect>>,
```

Comment:
```suggestion
    pub reflect_mut: for<'r, 'w> fn(&'r mut EntityMut<'w>) -> Option<Mut<'r, dyn Reflect>>,
```
Should split the lifetimes or this might lead to obscure but very annoying lifetime issues (afaik).

Reply:
Doesn't `'w` then become redundant?

Reply:
Maybe. I just make a habit of always tracking lifetimes accurately to the real things they're borrowing from, because otherwise it seems like I always run into weird situations where it's easy to have valid but not compiling borrows.

Reply:
`for<'a> fn(&'a mut EntityMut'a>)` is almost certainly _not_ what is wanted. Generally speaking you don't want to see `&'a mut T<'a>` since it requires the pointee to be borrowed for its entire life since unlike `&'a T<'a>` you cant subtype `&'a mut T<'long>` to `&'a mut T<'a>`.

`for<'a> fn(&'w mut EntityMut<'_>) -> Option<Mut<'a, dyn Reflect>>` is likely what is desired.

Reply:
> The returned lifetime will be the shorter one of 'r and 'w anyways.

as the type is `&'r mut EntityMut<'w>` the shorter one is _always_ going to be `'r` (or otherwise `'w` would be equal to `'r`). This is because in order for the type `&'a T` to be valid `T: 'a` has to hold, whichfor the case of `EntityMut<'w>: 'a` would require `'w: 'a`

Reply:
Just for my understanding; `for<'a> fn(&'w mut EntityMut<'_>) -> Option<Mut<'a, dyn Reflect>>` is equivalent but more concise than my suggestion, right?

Reply:
Yeah it's equivalent to your suggestion just more concise, I could have phrased it better. and yeah the `'w` was a typo, meant `'a` its not been updated thanks :)

Reply:
Changed to `for<'a> fn(&'w mut EntityMut) -> Option<Mut<'a, dyn Reflect>>`, is this ok or do I have to explicitly set `EntityMut<'_>`?

Reply:
Fixed :)